packages/providers/src/fuel-graphql-subscriber.ts

```diff
@@ -0,0 +1,75 @@
+import { FuelError } from '@fuel-ts/errors';
+import type { DocumentNode } from 'graphql';
+import { print } from 'graphql';
+
+type FuelGraphQLSubscriberOptions = {
+  url: string;
+  query: DocumentNode;
+  variables?: Record<string, unknown>;
+  fetchFn: typeof fetch;
+  abortController?: AbortController;
+};
+
+class FuelSubscriptionStream implements TransformStream {
+  readable: ReadableStream<FuelError | Record<string, unknown>>;
+  writable: WritableStream<Uint8Array>;
+  private readableStreamController!: ReadableStreamController<FuelError | Record<string, unknown>>;
+  private static textDecoder = new TextDecoder();
+
+  constructor() {
+    this.readable = new ReadableStream({
+      start: (controller) => {
+        this.readableStreamController = controller;
+      },
+    });
+
+    this.writable = new WritableStream<Uint8Array>({
+      write: (bytes) => {
+        const text = FuelSubscriptionStream.textDecoder.decode(bytes);
+        // the fuel node sends keep-alive messages that should be ignored
+        if (text.startsWith('data:')) {
+          const { data, errors } = JSON.parse(text.split('data:')[1]);
+          if (Array.isArray(errors)) {
+            this.readableStreamController.enqueue(
+              new FuelError(
+                FuelError.CODES.INVALID_REQUEST,
+                errors.map((err) => err.message).join('\n\n')
+              )
+            );
+          } else this.readableStreamController.enqueue(data);
+        }
+      },
+    });
+  }
+}
+
+export async function* fuelGraphQLSubscriber({
+  url,
+  variables,
+  query,
+  fetchFn,
+}: FuelGraphQLSubscriberOptions) {
+  const response = await fetchFn(`${url}-sub`, {
+    method: 'POST',
+    body: JSON.stringify({
+      query: print(query),
+      variables,
+    }),
+    headers: {
+      'Content-Type': 'application/json',
+      Accept: 'text/event-stream',
+    },
+  });
+
+  // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+  const subscriptionsStreamReader = response
+    .body!.pipeThrough(new FuelSubscriptionStream())
+    .getReader();
+
+  for (;;) {
+    const { value, done } = await subscriptionsStreamReader.read();
+    if (value instanceof FuelError) throw value;
+    yield value;
```

Comment:
Because it's an iterator, the fetch above won't run every time a new iteration occurs. Instead, once `yield` is hit, the next iteration will continue from the previously-hit `yield`, which means that it'll continue looping until `done` is `true`.

Reply:
Because a `while` loop would require an external variable to break it:

```ts
let doneStreaming = false;

while(!doneStreaming) {
    const { value, done } = await subscriptionStreamReader.read();
    if (value instanceof FuelError) throw value;
    yield value;
    doneStreaming = done;
}
```

I like breaking an infinite loop with `break` more in cases where the additional variable is created only for the purpose of breaking the loop and doesn't bring any additional context with it.

Reply:
Hmm, I have to agree here as well.

A more conventional syntax feels proper and keeps style consistency.

```ts
while (true) {
    const { value, done } = await subscriptionStreamReader.read();

    if (value instanceof FuelError) {
      throw value;
    }

    yield value;

    if (done) {
      break;
    }
}
```

Reply:
Ha! I didn't even consider breaking a while loop with ...`break`, lol.